src/routes/users.rs

```diff
@@ -54,6 +92,160 @@ async fn post_login(
     Ok(Redirect::to(&redirect_to).into_response())
 }
 
+async fn get_login_google(
+    State(state): State<AppState>,
+    session: Session,
+) -> ResponseResult<Response> {
+
+    let google_client_id = ClientId::new(
+        state.oauth_google_client_id
+    );
+    let google_client_secret = ClientSecret::new(
+        state.oauth_google_client_secret
+    );
+    let issuer_url = 
+        IssuerUrl::new("https://accounts.google.com".to_string()).context("failed to reach issuer")?;
+
+    let provider_metadata = task::spawn_blocking(move || {
+        GoogleProviderMetadata::discover(&issuer_url, http_client)
+        .context("failed to discover provider").unwrap()
+    }).await.context("failed to discover provider")?;
+    
+    let revocation_endpoint = provider_metadata
+        .additional_metadata()
+        .revocation_endpoint
+        .clone();
+    // Set up the config for the Google OAuth2 process.
+    let client = CoreClient::from_provider_metadata(
+        provider_metadata,
+        google_client_id,
+        Some(google_client_secret),
+    )
+    .set_redirect_uri(
+        RedirectUrl::new(state.base_url.to_string() + "/login_google_handler").expect("Invalid redirect URL"),
+    )
+    .set_revocation_uri(
+        RevocationUrl::new(revocation_endpoint).expect("Invalid revocation endpoint URL"),
+    );
+
+    // Generate the authorization URL to which we'll redirect the user.
+    let (authorize_url, csrf_state, nonce) = client
+        .authorize_url(
+            AuthenticationFlow::<CoreResponseType>::AuthorizationCode,
+            CsrfToken::new_random,
+            Nonce::new_random,
+        )
+        .add_scope(Scope::new("email".to_string()))
+        .add_scope(Scope::new("profile".to_string())).url();
+
+    // TODO: Store the CSRF and none states in a way that is more secure than this, although the current method is already quire secure.
+
+    session.insert("google_oidc_session", OidcSession { nonce, csrf_token:csrf_state }).await.context("failed to insert session")?;
+
+
+    Ok(Redirect::to(authorize_url.as_str()).into_response())
+}
+
+#[derive(Deserialize)]
+struct OAuthLoginQuery{
+    code: String,
+    state: String,
+}
+
+async fn get_login_google_handler(
+    session: Session,
+    Query(query): Query<OAuthLoginQuery>,
+    state: State<AppState>,
+    extract::Tx(mut tx): extract::Tx,
+) -> ResponseResult<Response> {
+    // get the nonce and csrf token from session
+    let oidc_session: OidcSession = session.get("google_oidc_session").await.context("failed to get session")?.context("session not found")?;
+
+    let code = AuthorizationCode::new(query.code.clone());
+    let csrf_state = CsrfToken::new(query.state.clone());
+
+    if csrf_state.secret() != oidc_session.csrf_token.secret() {
+        return Ok("Invalid CSRF token".into_response());
```

Comment:
This should never happen in a normal application. We don't need to tell users about the cause of the error, because they can do nothing to change it. As such, a normal `http 500` error is best:
```rs
return anyhow!("Invalid CSRF token");
```